### ISSUE TITLE
Non-dynamic profiling is not thread-safe

### ISSUE THREAD

**Issue by awkay** [ptaoussanis/tufte#64]
The logic in t.t.impl on line 83, which uses local pdata in non-dynamic mode isn't thread safe because of the use of a shared Stack that is globally declared. A combination of threads calling profile rapidly can lead to a situation where the modifications of a push or pop loses a stack entry, and the check of `empty` returns false when another thread has actually just corrupted and emptied the stack, leading to a stack empty exception.

Perhaps change to using https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ConcurrentLinkedDeque.html

**Comment by ptaoussanis** [comment]
@awkay Hi Tony, good catch- thanks for the report! PR welcome, otherwise I'll try look at this in the coming week.

Cheers!

**Comment by ptaoussanis** [comment]
Update: I should have a chance to get a fix up tomorrow.
Want to clean up a couple other things while I'm at it.

**Comment by awkay** [comment]
No hurry from my end (:dynamic? true is thread safe, so that works), and sorry I don't have time for a PR or I would. In the middle of a mess on other things :P

**Comment by ptaoussanis** [comment]
No problem, taking a closer look at this now.

> A combination of threads calling profile rapidly can lead to a situation where the modifications of a push or pop loses a stack entry, and the check of empty returns false when another thread has actually just corrupted and emptied the stack, leading to a stack empty exception.

Can you please confirm: you actually saw a stack empty exception?
`java.util.Stack` appears to be synchronized - and currently struggling to reproduce an exception.

**Comment by awkay** [comment]
Yes, I definitely was getting Empty stack exceptions. It's why I looked into it. All that has to happen is two pushes that interleave, where the second causes the first to be "missed". The structure will then cause two pops where there is only one thing on the stack. Of course, the probability is low, but you call it fast enough and it happens.

I'll review your fix and see what I think. multithreaded code is fun
